### PR TITLE
Revert "Reinstate load-time Pkg.precompile"

### PR DESCRIPTION
**Pull request by vchuravy**
Reverts JuliaLang/julia#51672

I think as is this PR can cause an infinite precompilation loop. When Pkg is not precompiled, requiring it will precompile Pkg, which will try to precompile Pkg.

### CODE REVIEW

**Comment by IanButterworth** [comment]
Revert sounds good but I don't follow that explanation. 

Also the PR was green before merge. Is something different about master runs than PR runs?

**Comment by IanButterworth** [comment]
Hold off on merging this. It could be a pkgserver issue

**Comment by KristofferC** [comment]
> When Pkg is not precompiled, requiring it will precompile Pkg, which will try to precompile Pkg.

So this does not happen?

**Comment by IanButterworth** [comment]
I was wrong (and confused by all the pkgserver stuff)

```
julia> using Pkg
ERROR: ConencyViolationError("deadlock detected in loading Pkg -> Pkg")
Stacktrace:
  [1] start_loading(modkey::Base.PkgId)
    @ Base ./loading.jl:1638
  [2] _require(pkg::Base.PkgId, env::Nothing)
    @ Base ./loading.jl:1965
  [3] __require_prelocked(uuidkey::Base.PkgId, env::Nothing)
    @ Base ./loading.jl:1863
  [4] #invoke_in_world#3
    @ Base ./essentials.jl:933 [inlined]
  [5] invoke_in_world
    @ Base ./essentials.jl:930 [inlined]
  [6] _require_prelocked
    @ Base ./loading.jl:1854 [inlined]
  [7] _require_prelocked
    @ Base ./loading.jl:1853 [inlined]
  [8] macro expansion
    @ Base ./lock.jl:267 [inlined]
  [9] require(uuidkey::Base.PkgId)
    @ Base ./loading.jl:1848
 [10] _require(pkg::Base.PkgId, env::String)
    @ Base ./loading.jl:2012
 [11] __require_prelocked(uuidkey::Base.PkgId, env::String)
    @ Base ./loading.jl:1863
 [12] #invoke_in_world#3
    @ Base ./essentials.jl:933 [inlined]
 [13] invoke_in_world
    @ Base ./essentials.jl:930 [inlined]
 [14] _require_prelocked(uuidkey::Base.PkgId, env::String)
    @ Base ./loading.jl:1854
 [15] macro expansion
    @ Base ./loading.jl:1792 [inlined]
 [16] macro expansion
    @ Base ./lock.jl:267 [inlined]
 [17] __require(into::Module, mod::Symbol)
    @ Base ./loading.jl:1754
 [18] #invoke_in_world#3
    @ Base ./essentials.jl:933 [inlined]
 [19] invoke_in_world
    @ Base ./essentials.jl:930 [inlined]
 [20] require(into::Module, mod::Symbol)
    @ Base ./loading.jl:1747

```

**Review by vtjnash — APPROVED** [review]

**Comment by Keno** [comment]
Why can't we just disable this while inside a precompile process? Unlike regular precompile, the Pkg-parallel-precompile code does not execute itself recursively, so it doesn't need to be loaded inside a precompile process at all.

**Comment by KristofferC** [comment]
I also asked this here https://github.com/JuliaLang/julia/pull/51676#issuecomment-1759657818.

**Comment by IanButterworth** [comment]
For the record, it already did that. Note the `!generating_output `check. 

https://github.com/JuliaLang/julia/pull/51676 seemed like the fix to me but I was unaware of the load graph issues Jameson raised in there. 

Given the whole of `Pkg.precompile` cannot move to Base (it resolves and instantiates first) it does make me think that this might just have be a feature that isn't enabled until Pkg has been loaded elsewhere (by a repl switch, etc).

**Comment by vtjnash** [comment]
Right, this hook was fundamentally changing the way that `require` works, but only when Pkg is loaded (for example, it has the ability to change the set of packages that are being loaded into a different set that are incompatible with the current set that was partially loaded already).

It can't be disabled only in the precompile process because it the failure happens in the driver process, where we have already acquired a lock on the precompile of child task, when we discover that Pkg is desired, causing an edge to be added from that child task to Pkg. If that child happens to also be a dependency of Pkg, that operation causes a deadlock and the observed crashing.

**Comment by Keno** [comment]
So just hardcode the list of Pkg's dependencies?

**Comment by Keno** [comment]
Do we have a plan here? This is a bit of a trap for people upgrading to master who have become accustomed to automatic parallel precompile (esp since the serial precompile fallback is silent).

**Comment by IanButterworth** [comment]
> esp since the serial precompile fallback is silent

Just clarifying, it's an `@info` in interactive mode, `@debug` in noninteractive.

A workaround for the user is to put `import Pkg` in their startup.jl, which will load the hook, or just let people know they need to switch to `pkg>` mode once to enable it.

For the plan to reinstate the behavior from launch, I don't know what to suggest. IMO, the main issue is actually that we need to call `Pkg.instantiate` from Base (which `Pkg.precompile` calls first). Moving the parallel precompilation part of `Pkg.precompile` to Base is doable otherwise.

**Comment by Keno** [comment]
> Just clarifying, the old serial precompile messaging is an `@info` in interactive mode, `@debug` in noninteractive mode.

Yes, I figured somebody just broke the printing when I saw that.

> For the plan to reinstate the behavior from launch, I don't know what to suggest. IMO, the main issue is actually that we need to call Pkg.instantiate from Base (which Pkg.precompile calls first). Moving the parallel precompilation part of Pkg.precompile to Base is doable otherwise.

Why can't we just do the dependency hardcoding thing for now? It might be ugly, but it'll work at least or am I missing something?

**Pull request merged**
